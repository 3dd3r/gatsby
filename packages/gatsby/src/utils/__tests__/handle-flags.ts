import handleFlags from "../handle-flags"
import _ from "lodash"
import { IFlag } from "./flags"

jest.mock(`gatsby-core-utils`, () => {
  return {
    isCI: (): boolean => true,
  }
})

describe(`handle flags`, () => {
  const activeFlags: Array<IFlag> = [
    {
      name: `FAST_DEV`,
      env: `GATSBY_EXPERIMENTAL_FAST_DEV`,
      command: `develop`,
      description: `Enable all experiments aimed at improving develop server start time`,
      includedFlags: [`DEV_SSR`, `QUERY_ON_DEMAND`],
    },
    {
      name: `DEV_SSR`,
      env: `GATSBY_EXPERIMENTAL_DEV_SSR`,
      command: `develop`,
      description: `SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.`,
      umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/28138`,
    },
    {
      name: `QUERY_ON_DEMAND`,
      env: `GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND`,
      command: `develop`,
      description: `Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.`,
      umbrellaIssue: `https://github.com/gatsbyjs/gatsby/discussions/27620`,
      noCI: true,
    },
    {
      name: `ONLY_BUILDS`,
      env: `GATSBY_EXPERIMENTAL_BUILD_ME_FASTER`,
      command: `build`,
      description: `test`,
      umbrellaIssue: `test`,
    },
    {
      name: `ALL_COMMANDS`,
      env: `GATSBY_EXPERIMENTAL_SOMETHING_COOL`,
      command: `all`,
      description: `test`,
      umbrellaIssue: `test`,
    },
    {
      name: `YET_ANOTHER`,
      env: `GATSBY_EXPERIMENTAL_SOMETHING_COOL2`,
      command: `all`,
      description: `test`,
      umbrellaIssue: `test`,
    },
    {
      name: `PARTIAL_RELEASE`,
      env: `GATSBY_READY_TO_GO`,
      command: `all`,
      description: `test`,
      umbrellaIssue: `test`,
      partialRelease: {
        percentage: 100,
        releaseNotes: `https://example.com`,
      },
    },
    {
      name: `PARTIAL_RELEASE_ONLY_VERY_OLD_LODASH`,
      env: `GATSBY_READY_TO_GO_LODASH`,
      command: `all`,
      description: `test`,
      umbrellaIssue: `test`,
      semver: {
        // Because of this, this flag will never show up
        lodash: `<=3.9`,
      },
      partialRelease: {
        percentage: 100,
        releaseNotes: `https://example.com`,
      },
    },
    {
      name: `PARTIAL_RELEASE_ONLY_NEW_LODASH`,
      env: `GATSBY_READY_TO_GO_NEW_LODASH`,
      command: `all`,
      description: `test`,
      umbrellaIssue: `test`,
      semver: {
        lodash: `>=4.9`,
      },
      partialRelease: {
        percentage: 100,
        releaseNotes: `https://example.com`,
      },
    },
  ]

  const configFlags = {
    FAST_DEV: true,
    ALL_COMMANDS: true,
  }
  const configFlagsWithFalse = {
    ALL_COMMANDS: true,
    DEV_SSR: false,
  }
  const configWithFlagsNoCi = {
    QUERY_ON_DEMAND: true,
    DEV_SSR: true,
  }

  it(`returns enabledConfigFlags and a message`, () => {
    expect(handleFlags(activeFlags, configFlags, `develop`)).toMatchSnapshot()
  })

  it(`filters out flags marked false`, () => {
    expect(
      handleFlags(activeFlags, configFlagsWithFalse, `develop`)
        .enabledConfigFlags
    ).toHaveLength(3)
  })

  it(`filters out flags that are marked as not available on CI`, () => {
    expect(
      handleFlags(activeFlags, configWithFlagsNoCi, `develop`)
        .enabledConfigFlags
    ).toHaveLength(3)
  })

  it(`filters out flags that aren't for the current command`, () => {
    expect(
      handleFlags(activeFlags, configFlags, `build`).enabledConfigFlags
    ).toHaveLength(3)
  })

  it(`returns a message about unknown flags in the config`, () => {
    const unknownConfigFlags = handleFlags(
      activeFlags,
      { ALL_COMMANDS: true, FASTLY_DEV: true, SUPER_COOL_FLAG: true },
      `develop`
    )
    expect(unknownConfigFlags).toMatchSnapshot()
  })

  it(`opts in sites to a flag if their site is selected for partial release`, () => {
    // Nothing is enabled in their config.
    const response = handleFlags(activeFlags, {}, `develop`)
    expect(response).toMatchSnapshot()
  })

  it(`removes flags people explicitly opt out of and ignores flags that don't pass semver`, () => {
    const response = handleFlags(
      activeFlags,
      { PARTIAL_RELEASE: false, PARTIAL_RELEASE_ONLY_NEW_LODASH: false },
      `develop`
    )
    expect(response.enabledConfigFlags).toHaveLength(0)
  })
})
