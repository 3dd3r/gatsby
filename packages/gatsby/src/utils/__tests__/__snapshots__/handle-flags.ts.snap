// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handle flags opts in sites to a flag if their site is selected for partial release 1`] = `
Object {
  "enabledConfigFlags": Array [
    Object {
      "command": "all",
      "description": "test",
      "env": "GATSBY_READY_TO_GO",
      "name": "PARTIAL_RELEASE",
      "partialRelease": Object {
        "percentage": 100,
        "releaseNotes": "https://example.com",
      },
      "umbrellaIssue": "test",
    },
  ],
  "message": "The following flags are active:
We're shipping new features! For final testing, we're rolling them out first to a small % of Gatsby users
and your site was automatically choosen as one of them. With your help, we'll then release them to everyone in the next minor release

We greatly appreciate your help testing the change. Please report any feedback good or bad in the umbrella issue. If you do encounter problems, please disable the flag by setting it to false in your gatsby-config.js like:

flags: {
  THE_FLAG: false
}

The following flags are enabled on your site:
- PARTIAL_RELEASE · (Umbrella Issue (test)) · test

There are 6 other flags available that you might be interested in:
- FAST_DEV · Enable all experiments aimed at improving develop server start time
- DEV_SSR · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/28138)) · SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.
- QUERY_ON_DEMAND · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/27620)) · Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.
- ONLY_BUILDS · (Umbrella Issue (test)) · test
- ALL_COMMANDS · (Umbrella Issue (test)) · test
- YET_ANOTHER · (Umbrella Issue (test)) · test
",
  "unknownFlagMessage": "",
}
`;

exports[`handle flags returns a message about unknown flags in the config 1`] = `
Object {
  "enabledConfigFlags": Array [
    Object {
      "command": "all",
      "description": "test",
      "env": "GATSBY_EXPERIMENTAL_SOMETHING_COOL",
      "name": "ALL_COMMANDS",
      "umbrellaIssue": "test",
    },
    Object {
      "command": "all",
      "description": "test",
      "env": "GATSBY_READY_TO_GO",
      "name": "PARTIAL_RELEASE",
      "partialRelease": Object {
        "percentage": 100,
        "releaseNotes": "https://example.com",
      },
      "umbrellaIssue": "test",
    },
  ],
  "message": "The following flags are active:
- ALL_COMMANDS · (Umbrella Issue (test)) · test
We're shipping new features! For final testing, we're rolling them out first to a small % of Gatsby users
and your site was automatically choosen as one of them. With your help, we'll then release them to everyone in the next minor release

We greatly appreciate your help testing the change. Please report any feedback good or bad in the umbrella issue. If you do encounter problems, please disable the flag by setting it to false in your gatsby-config.js like:

flags: {
  THE_FLAG: false
}

The following flags are enabled on your site:
- PARTIAL_RELEASE · (Umbrella Issue (test)) · test

There are 5 other flags available that you might be interested in:
- FAST_DEV · Enable all experiments aimed at improving develop server start time
- DEV_SSR · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/28138)) · SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.
- QUERY_ON_DEMAND · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/27620)) · Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.
- ONLY_BUILDS · (Umbrella Issue (test)) · test
- YET_ANOTHER · (Umbrella Issue (test)) · test
",
  "unknownFlagMessage": "The following flag(s) found in your gatsby-config.js are not known:
- FASTLY_DEV (did you mean: FAST_DEV)
- SUPER_COOL_FLAG",
}
`;

exports[`handle flags returns validConfigFlags and a message 1`] = `
Object {
  "enabledConfigFlags": Array [
    Object {
      "command": "develop",
      "description": "Enable all experiments aimed at improving develop server start time",
      "env": "GATSBY_EXPERIMENTAL_FAST_DEV",
      "includedFlags": Array [
        "DEV_SSR",
        "QUERY_ON_DEMAND",
      ],
      "name": "FAST_DEV",
    },
    Object {
      "command": "all",
      "description": "test",
      "env": "GATSBY_EXPERIMENTAL_SOMETHING_COOL",
      "name": "ALL_COMMANDS",
      "umbrellaIssue": "test",
    },
    Object {
      "command": "all",
      "description": "test",
      "env": "GATSBY_READY_TO_GO",
      "name": "PARTIAL_RELEASE",
      "partialRelease": Object {
        "percentage": 100,
        "releaseNotes": "https://example.com",
      },
      "umbrellaIssue": "test",
    },
    Object {
      "command": "develop",
      "description": "SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.",
      "env": "GATSBY_EXPERIMENTAL_DEV_SSR",
      "name": "DEV_SSR",
      "umbrellaIssue": "https://github.com/gatsbyjs/gatsby/discussions/28138",
    },
    Object {
      "command": "develop",
      "description": "Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.",
      "env": "GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND",
      "name": "QUERY_ON_DEMAND",
      "noCI": true,
      "umbrellaIssue": "https://github.com/gatsbyjs/gatsby/discussions/27620",
    },
  ],
  "message": "The following flags are active:
- FAST_DEV · Enable all experiments aimed at improving develop server start time
- ALL_COMMANDS · (Umbrella Issue (test)) · test
- DEV_SSR · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/28138)) · SSR pages on full reloads during develop. Helps you detect SSR bugs and fix them without needing to do full builds.
- QUERY_ON_DEMAND · (Umbrella Issue (https://github.com/gatsbyjs/gatsby/discussions/27620)) · Only run queries when needed instead of running all queries upfront. Speeds starting the develop server.
We're shipping new features! For final testing, we're rolling them out first to a small % of Gatsby users
and your site was automatically choosen as one of them. With your help, we'll then release them to everyone in the next minor release

We greatly appreciate your help testing the change. Please report any feedback good or bad in the umbrella issue. If you do encounter problems, please disable the flag by setting it to false in your gatsby-config.js like:

flags: {
  THE_FLAG: false
}

The following flags are enabled on your site:
- PARTIAL_RELEASE · (Umbrella Issue (test)) · test

There are 2 other flags available that you might be interested in:
- ONLY_BUILDS · (Umbrella Issue (test)) · test
- YET_ANOTHER · (Umbrella Issue (test)) · test
",
  "unknownFlagMessage": "",
}
`;
